### PR TITLE
feat: multi-toolchain support

### PR DESCRIPTION
**Pull request by f0rmiga**
## PR Checklist

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Most mono repos using Bazel have the problem of needing support for multiple Python versions. Currently, there's no simple way to accomplish this.

## What is the new behavior?

Add support for registering multiple Python toolchains with different versions as an opt-in feature so that users can gradually migrate portions of their mono repos to this new feature.

Using load statements to override `py_binary` and `py_test`, this feature implementation also plays well with the Gazelle extension by using Gazelle's `# gazelle:map_kind from_kind to_kind to_kind_load` feature. Also, pure Python libraries can take advantage of this by testing against multiple Python versions in the same BUILD file (see examples).

This feature is achieved by using platform transitions behind the scenes.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

File: python/platforms/transition.bzl
Comment (rickeylev) [inline python/platforms/transition.bzl:111]: Lets move the list of these attribute names into a separate file. It will most assuredly get reused by other code.

File: python/platforms/transition.bzl
Comment (rickeylev) [inline python/platforms/transition.bzl:126]: The "name" arg is being removed in an upcoming release; it ended up causing some sort of cache poisoning type of issue or something.

You'll need to change this to bind the return value to an global variable.

File: python/platforms/transition.bzl
Comment (rickeylev) [inline python/platforms/transition.bzl:158]: It seems like there's a lot of copy/paste for these common attributes. How about instead of manually carrying around ~11 variables, we stick them in a dict? 11 is a lot to carry around and makes it easy to forget one.

File: python/repositories.bzl
Comment (rickeylev) [inline python/repositories.bzl:398]: Why does this need to be an opt-in setting instead of on by default?

File: python/repositories.bzl
Comment (rickeylev) [inline python/repositories.bzl:475]: Can you explain this a bit more? I don't fully understand the problem.

Each of the registered toolchains should have e.g. compatible_with=[X] such that they're mutually exclusive. When a rule does toolchain lookup, the rule should already have a python version set so that when Bazel matches its config to the available toolchains, it can only match one.

It sounds like the either the "mutually exclusive" part or the "rule's setting is set" part isn't true, and multiple toolchains are matching?

File: python/platforms/transition.bzl
Comment (rickeylev) [inline python/platforms/transition.bzl:129]: Also pass along env_inherit.

File: python/platforms/transition.bzl
Comment (rickeylev) [inline python/platforms/transition.bzl:125]: I think you also want to pass along "srcs" and "deps" here. They can also participate in expansion for args and env.

**Review by rickeylev — COMMENTED** [review]
I'm not sure about the usage of platforms here. It seems like too coarse an object. All that's is being varied is the python version, but a platform encompasses a lot more than that.

I guess, stepping back, what's the usage of this look like and use case?

When you mentioned it in chat, I thought you meant something like:

```
py_binary(name="a", python_version="3.8", data=[":b"], deps=[":c"])
py_binary(name="b", python_version="3.9", deps=[":c"])
whatever(name="c", srcs = select({
  "py3.8": ...,
  "py3.9": ...
}))
```

And the end result is two runtimes end up in runfiles, and each respective binary uses its associated interpreter, and dependencies are able to just naturally select on the python version.

But, I see platforms here, and there's many edits to the pip stuff, so I'm not sure if what I expected is entirely accurate.

**Comment by f0rmiga** [comment]
> I'm not sure about the usage of platforms here. It seems like too coarse an object. All that's is being varied is the python version, but a platform encompasses a lot more than that.

We definitely want to use platforms here. While we can't yet do cross-platform builds with native wheels, we can support this soon (a common case being macos -> linux containers).

> I guess, stepping back, what's the usage of this look like and use case?

Commonly, a mono repo has multiple teams and projects, leading to different Python version requirements. This allows projects to move to newer Python versions independently.

> When you mentioned it in chat, I thought you meant something like:
> 
> ```
> py_binary(name="a", python_version="3.8", data=[":b"], deps=[":c"])
> py_binary(name="b", python_version="3.9", deps=[":c"])
> whatever(name="c", srcs = select({
>   "py3.8": ...,
>   "py3.9": ...
> }))
> ```

Why would someone want to deal with selects manually? Transitions are here to solve exactly this problem.

> And the end result is two runtimes end up in runfiles, and each respective binary uses its associated interpreter, and dependencies are able to just naturally select on the python version.

That's exactly the case with my implementation. See the tests.

> But, I see platforms here, and there's many edits to the pip stuff, so I'm not sure if what I expected is entirely accurate.

Your expectation and the community's expectations may not be aligned on this.

File: examples/multi_python_versions/.bazelrc
Comment (f0rmiga) [reply examples/multi_python_versions/.bazelrc:3]: This is a good question. I don't know many people developing with Windows. @groodt What do you think?

File: python/platforms/transition.bzl
Comment (f0rmiga) [reply python/platforms/transition.bzl:126]: Could you point me to some docs or PRs talking about this, please?

File: python/repositories.bzl
Comment (f0rmiga) [reply python/repositories.bzl:398]: This allows a default toolchain to be picked when multi-python support is not enabled.

File: examples/multi_python_versions/.bazelrc
Comment (alexeagle) [reply examples/multi_python_versions/.bazelrc:3]: Not that it's necessarily correct, but rules_js has a hard requirement of enabling runfiles on Windows and I don't know of any complaints we've gotten from Windows developers who aren't able to use it.

**Comment by rickeylev** [comment]
So, keep in mind I'm coming into this without much context. A lot of this code is new to me, and I only know the broad strokes of how the repository rules like requirements, pip parse, and the standalone interpreter rules work. I'm happy to VC to chat about some the points in this PR.

> Commonly, a mono repo has multiple teams and projects, leading to different Python version requirements. This allows projects to move to newer Python versions independently.

This basically sounds like the python 2/3 situation (i.e. the `py_binary.python_version` attribute and --python_version flag), except applied to 3.X value. Is that correct?

That, too uses a transition, however, it has two key differences:
 * That transition only affects the python version, not the entire platform.
 * The python version is a parameter of the rule, rather than a rule itself.

The first means any surrounding platform configuration is inherited. This essentially allows an easy way to have a single target definition that can be built for multiple platforms.

The second means only a single rule is necessary, rather than N rules. It also means a default can be obtained from elsewhere rather than it _always_ being forced.

> 
> > When you mentioned it in chat, I thought you meant something like:
> > ```
> > py_binary(name="a", python_version="3.8", data=[":b"], deps=[":c"])
> > py_binary(name="b", python_version="3.9", deps=[":c"])
> > whatever(name="c", srcs = select({
> >   "py3.8": ...,
> >   "py3.9": ...
> > }))
> > ```
> 
> Why would someone want to deal with selects manually? Transitions are here to solve exactly this problem.

Selects and transitions solve different but related problems. Transitions are for forcing a transitive closure to have a particular configuration. Selects are for selecting different dependencies/inputs based upon the currently active configuration.

For the problem of: "I need to use a different target depending on the python version, OS, cpu, etc", then the answer is to use select. Whether it's a wheel, py_library, filegroup, or what not, doesn't particularly matter. Usually the consumed target hides this detail from the consumed target by using alias() and/or varying the inputs to its canonical target.

As an example, lets say we wanted to build an binary using importlib resources for multiple Python versions. I use this example because importlib_resources (underscore) is a backport of functionality that is in the stdlib later. At the same time, the contents of the importlib_resources wheel could contain Python version specific artifacts (e.g. pyc files). To make this work, you would end up with something like this:

```
# myapp/BUILD
py_binary(
  name="myapp",
  deps = select({
    "py3.12": [] # 3.12 finally has all the features we need; no dep needed
    "default": [requirements("importlib_resources")]
  })
)

# importlib_resources/BUILD
# I forget the exact generated BUILD file contents, but I hope this
# gets the gist across.
# Method 1: a single entry point whose inputs are varied. Makes sense when
# the variation is just on some of the attributes.
py_wheel(
  name = "importlib_resources",
  srcs = select({
    "py3.7": ["importlib_resources-py37.wheel"],
    "py3.8": ["importlib_resources-py38.wheel"],
    "py3.9": ...
    "py3.10": ...
    "py3.11": ...
    "py3.12": ...
  })
)
# Method 2: an alias that dispatches to the appropriate target. Makes sense
# when the variation is more nuanced.
alias(
  name = "importlib_resources",
  actual = select({
    "py3.8": ":_importlib_resources_38",
    "py3.9": ":_importlib_resources_39",
    ...
}))
py_library(name="_importlib_resources_{XX}", ...) # etc
```

The above is mostly orthogonal to a transition. The transition just changes where the active configuration is being set. Note I'm just a simple first-order dependency relationship for clarity. The path from "myapp" to "importlib_resources" could have an arbitrary number of intervening dependencies, especially for a monorepo that code shares freely.

Another chunk I don't understand `compile_pip_requirements_X_Y` -- why must there be a separate rule for each version? I think it can be replaced with e.g.
```
compile_pip_requirements(
  name = "requirements",
  requirements_txt = select({"py38": "requirements_38.txt", "py39": "requirements_39.txt", etc})
)
```

And if the version does needs to be forced, an attribute can specify that.

Similarly, `multi_pip_parse` takes version strings as inputs to map the versions to interpreters -- this looks suspiciously like the sort of thing that toolchain resolution is supposed to handle? e.g. a toolchain is registered to be compatible for python 3.8, and then pip_parse obtains the interpreter from the toolchain, and finding the matching interpreter for the current configuration is handled by toolchain resolution.

For the problem of: "I need this target to build with a specific configuration", then the answer is to use a transition.

To be clear, yes, a transition definitely needs to be involved here as, at a minimum, the python version must be changed to affect the transitive closure. My issue isn't with using a transition, but that the platform needs to be transitioned instead of simply the python version.

As an example, if I'm understanding this code correctly, this looks to be possible, but probably not intended:

```
py_binary(
  name="outer", target_platform="@rules_python//:darwin_py38",
  data = [":inner"],
)
py_binary(
  name = "inner", target_platform="@rules_python//:linux_py38"
)
```

The net effect is "outer" is built for darwin and includes the "inner" binary built for linux. This is probably not intended. Making this work requires inner to add a separate target with a platform-specific name and associated alias mapping for each platform a consumer needs, and also results in the consumer needing to handle the runfiles artifact name varying depending on the platform.

File: python/platforms/transition.bzl
Comment (rickeylev) [reply python/platforms/transition.bzl:126]: https://bazel.build/rules/lib/globals#rule (last arg in the args table).
https://github.com/bazelbuild/bazel/issues/16301

File: python/repositories.bzl
Comment (rickeylev) [reply python/repositories.bzl:475]: Hm. Ok. Why is there a toolchain without a version constraint? Do we not know the python version when the toolchain is registered?

File: examples/multi_python_versions/.bazelrc
Comment (rickeylev) [reply examples/multi_python_versions/.bazelrc:3]: That sounds very promising. I also found these:
https://github.com/bazelbuild/bazel/issues/5807
https://github.com/bazelbuild/bazel/issues/8704

In the reference history of the first one, it looks like several other rules also enabled symlinks for windows. The second appears to be a canonical "enable by default" bug. 

I also found this doc: https://bazel.build/configure/windows#symlink
Which seems to both recommend using and not using it :shrug:

**Comment by rickeylev** [comment]
Per chat:

re: specifying the python version using the rule name vs attribute: I'm fine with having e.g. `py_binary_3_8` etc; I don't have a strong opinion there. There's pros/cons to encoding this in the rule name vs attribute. If people see value in version-specific macros, then yeah, lets define those (Thulio mentioned something about how Gazelle can be told to rewrite the rule used, which definitely sounds like an easy and appealing way to bulk-update build files)

re: transitioning platforms: I asked some other Starlark experts, and they agreed transitioning platform is a pretty heavy weight way to do it. A better way is to create a flag that a transition affects and use config_setting to map the flag to constraints for the toolchains.

Putting it all together, I think it looks something like this:

```
# https://bazel.build/extending/config#predefined-settings
load("@bazel_skylib//rules:common_settings.bzl", "string_flag")
string_flag(name="python_version", default="XXX", values=["3.6", "3.7", ...)

config_setting(
  name = "is_py3.6",
  flag_values = {":python_version": "3.6"}
)
config_setting(
  name = "is_py3.7"
  flag_values = {":python_version": "3.7"}
)

py_runtime(
  name = "py3.6_runtime",
  target_compatible_with = [":is_py3.6"],
)
py_runtime(
  name = "py3.7_runtime",
  target_compatible_with = [":is_py3.7"],
)
toolchain(
  name = "py3.6_toolchain",
  target_compatible_with = [":is_py3.6", ...],
  toolchain = ":py3.6_runtime",
  toolchain_type = ...
)
toolchain(
  name = "py3.7_toolchain",
  target_compatible_with = [":is_py3.7", ...],
  toolchain = ":py3.7_runtime",
  toolchain_type = ...
)

def py_version_transition_impl(...):
  return {":python_version": ctx.attr.python_version}

def py_versioned_binary_impl(ctx):
  ...same as before...
py_versioned_binary = rule(cfg=py_version_transition, ...)
```

**Comment by f0rmiga** [comment]
@rickeylev

re: specifying the python version using the rule name vs attribute: you don't necessarily need to rename, or follow what I did; see below:

https://github.com/bazelbuild/rules_python/blob/1053ab06b16c7e346e8b6a253fd3c45bcc4e686d/examples/multi_python_versions/tests/BUILD.bazel#L1-L4

I can think of a few cases on a monorepo for doing the renaming. The most obvious would be a monorepo where a library is shared between multiple projects, and those projects end up having different python version requirements. The maintainers of the shared library will want to assert the tests pass on all python versions, so a rename like this would be required. Otherwise, it makes more sense to just change the import statement on a BUILD file and it's like magic to move to another version. E.g. when adopting this new feature:

```diff
-load("@rules_python//python:defs.bzl", "py_binary", "py_test")
+load("@python//3.8:defs.bzl", "py_binary", "py_test")
```

re: transitioning platforms: I pushed https://github.com/bazelbuild/rules_python/pull/846/commits/1053ab06b16c7e346e8b6a253fd3c45bcc4e686d with this. Indeed much simpler. Thanks!

File: python/platforms/transition.bzl
Comment (f0rmiga) [reply python/platforms/transition.bzl]: Addressed in d813740.

File: python/platforms/transition.bzl
Comment (f0rmiga) [reply python/platforms/transition.bzl]: Addressed in 161f068.

File: python/platforms/transition.bzl
Comment (f0rmiga) [reply python/platforms/transition.bzl]: I actually missed a few test and binary-specific attributes. Addressed in 66b33f3.

File: python/platforms/transition.bzl
Comment (f0rmiga) [reply python/platforms/transition.bzl]: Addressed in cecb3c8.

File: python/config_settings/transition.bzl
Comment (rickeylev) [inline python/config_settings/transition.bzl:53]: Ugh, this is annoying. Does checking hasattr(native, "RunEnvironmentInfo") work?

Otherwise lets mention this FR in the comment for now: https://github.com/bazelbuild/bazel/issues/8305

I also filed #901 for this.

iirc, TestEnvironment is an alias to RunEnvironmentInfo, so it's NBD really.

File: python/config_settings/transition.bzl
Comment (rickeylev) [inline python/config_settings/transition.bzl:77]: You'll want to apply the transition to the whole rule (`rule(cfg=...)`, otherwise things like deps, srcs, and tools won't select based upon the same version of python as what's in the `target` attribute.

File: python/config_settings/transition.bzl
Comment (rickeylev) [inline python/config_settings/transition.bzl:49]: To make sure the coverage instrumentation filters match the wrapped binary, specify the target attribute as one of the dependency attributes. The other attributes with the common names (srcs, deps, etc) should be automatically handled.
```suggestion
        # Ensure that the binary we're wrapping is included in code coverage.
        coverage_common.instrumented_files_info(
            ctx,
            dependency_attributes = ["target"],
        ),      
```

File: python/config_settings/transition.bzl
Comment (rickeylev) [inline python/config_settings/transition.bzl:37]: So I understand what this logic is doing, but why is it doing it? The underlying zip file will be there. Is it just to make sure that "{name}.zip" shows up alongside the regular "{name}" instead of the underlying "_{name}.zip"?

That's fine, but why does it matter? Please add a brief comment to explain.

It's unfortunate we have to flatten here; I don't see a way to avoid that. A minor improvement would be to flatten the default outputs instead of runfiles (it'd be a smaller depset); though I'm not sure how much it'll actually help because, under the hood, building the zip requires flattening anyways.

File: python/pip.bzl
Comment (rickeylev) [inline python/pip.bzl:168]: I don't understand the "why" of these pip/wheel changes, so lgtm!

File: python/config_settings/BUILD.bazel
Comment (rickeylev) [inline python/config_settings/BUILD.bazel:5]: What are these distribution targets?

**Review by rickeylev — CHANGES_REQUESTED** [review]
Mostly looks good! Ok to merge after addressing comments.

File: python/config_settings/BUILD.bazel
Comment (f0rmiga) [reply python/config_settings/BUILD.bazel:5]: Used to collect sources for bazel-in-bazel integration tests.

File: python/config_settings/transition.bzl
Comment (f0rmiga) [reply python/config_settings/transition.bzl:42]: Addressed in 906bb6a.

File: python/config_settings/transition.bzl
Comment (f0rmiga) [reply python/config_settings/transition.bzl]: Addressed in 9c833f1.

File: python/config_settings/transition.bzl
Comment (f0rmiga) [reply python/config_settings/transition.bzl]: Checking if `native` has it works, thanks for the suggestion. Addressed in aa0fc9f.

File: python/config_settings/transition.bzl
Comment (rickeylev) [inline python/config_settings/transition.bzl:84]: This needs to be cfg=target, otherwise it could build for a different platform that what it'll run on. e.g., with remote execution, it might build for linux, while locally you build and run a test for mac.

```suggestion
        cfg = "target",
```

**Review by rickeylev — CHANGES_REQUESTED** [review]

File: python/config_settings/transition.bzl
Comment (f0rmiga) [reply python/config_settings/transition.bzl]: Of course... addressed in 6896f15.

**Review by rickeylev — APPROVED** [review]

**Pull request merged**
